target/linux/ath79/dts/ar7161_meraki_mr16.dts

```diff
@@ -0,0 +1,178 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/dts-v1/;
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+#include "ar7100.dtsi"
+
+/ {
+	compatible = "meraki,mr16", "qca,ar7161";
+	model = "Meraki MR16";
+
+	chosen {
+		bootargs = "console=ttyS0,115200";
+	};
+
+	aliases {
+		led-boot = &power_green;
+		led-failsafe = &power_orange;
+		led-running = &power_green;
+		led-upgrade = &power_green;
+	};
+
+	extosc: ref {
+		compatible = "fixed-clock";
+		#clock-cells = <0>;
+		clock-output-names = "ref";
+		clock-frequency = <40000000>;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		wifi1 {
+			label = "mr16:green:wifi1";
+			gpios = <&gpio 0 GPIO_ACTIVE_LOW>;
+		};
+
+		wifi2 {
+			label = "mr16:green:wifi2";
+			gpios = <&gpio 1 GPIO_ACTIVE_LOW>;
+		};
+
+		wifi3 {
+			label = "mr16:green:wifi3";
+			gpios = <&gpio 2 GPIO_ACTIVE_LOW>;
+		};
+
+		wifi4 {
+			label = "mr16:green:wifi4";
+			gpios = <&gpio 3 GPIO_ACTIVE_LOW>;
+		};
+
+		wan {
+			label = "mr16:green:wan";
+			gpios = <&gpio 4 GPIO_ACTIVE_LOW>;
+		};
+
+		power_orange: power_orange {
+			label = "mr16:orange:power";
+			gpios = <&gpio 5 GPIO_ACTIVE_LOW>;
+			panic-indicator;
+		};
+
+		power_green: power_green {
+			label = "mr16:green:power";
+			gpios = <&gpio 6 GPIO_ACTIVE_LOW>;
+		};
+	};
+
+	keys {
+		compatible = "gpio-keys";
+
+		reset {
+			label = "reset";
+			linux,code = <KEY_RESTART>;
+			gpios = <&gpio 7 GPIO_ACTIVE_LOW>;
+		};
+	};
+};
+
+&pcie0 {
+	status = "okay";
+
+	ath9k0: wifi@0,11 { /* 2.4 GHz */
+		compatible = "pci168c,0029";
+		reg = <0x8800 0 0 0 0>;
+		qca,no-eeprom;
+		mtd-mac-address = <&mac 0x0>;
+		mtd-mac-address-increment = <1>;
+		#gpio-cells = <2>;
+		gpio-controller;
+	};
+
+	ath9k1: wifi@0,12 { /* 5 GHz */
+		compatible = "pci168c,0029";
+		reg = <0x9000 0 0 0 0>;
+		qca,no-eeprom;
+		mtd-mac-address = <&mac 0x0>;
+		mtd-mac-address-increment = <2>;
+		#gpio-cells = <2>;
+		gpio-controller;
+	};
+};
+
+&pll {
+	clocks = <&extosc>;
+};
+
+&uart {
+	status = "okay";
+};
+
+&mdio0 {
+	status = "okay";
+
+	phy-mask = <0x1>;
+
+	phy0: ethernet-phy@0 {
+		reg = <0>;
+	};
+};
+
+&eth0 {
+	status = "okay";
+	mtd-mac-address = <&mac 0x0>;
+
+	pll-data = <0x00110000 0x00001099 0x00991099>;
+
+	phy-mode = "rgmii";
+	phy-handle = <&phy0>;
+};
+
+&spi {
+	status = "okay";
+	num-cs = <1>;
+
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <25000000>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "u-boot";
+				reg = <0x000000 0x40000>;
+				read-only;
+			};
+
+			partition@40000 {
+				label = "u-boot-env";
+				reg = <0x40000 0x40000>;
+				read-only;
+			};
+
+			partition@80000 {
+				label = "firmware";
+				reg = <0x80000 0xf50000>;
+				compatible = "denx,uimage";
+			};
+
+			mac: partition@fd0000 {
+				label = "mac";
+				reg = <0xfd0000 0x10000>;
```

Comment:
missing readonly

Reply:
This must be written by the user as it is not contained in a partition in OEM. As best I understand at this point, the OEM config pulls it from the ART. I could supply programs to modify the ART, but that creates yet another problem.

Reply:
Well, if the OEM firmware puts it here it must get it from somewhere. I don't think reading from an undefined intermediate is desirable. We should read from the original source.

Reply:
And if the user has to put it there, he can as well put it in uci config.

Reply:
This makes for sticky situations when attempting to put routers into recovery mode. I agree about reading from source if possible. I will dig later for the original commit which did this; I suspect the source MAC is in the OEM configuration of the board.

That said -- for what it's worth, this is how the Openwrt 18.06 and onward MR16 is actually configured, meaning that we now have to roll backwards to find which of two locations is the correct one if a user is upgrading from 18.06.

Reply:
> in the OEM configuration of the board

-- to be clear, in the middle of the firmware partition, in the OEM's root filesystem.

I will dump from an unflashed Meraki and attempt to correspond it.

Reply:
> That said -- for what it's worth, this is how the Openwrt 18.06 and onward MR16 is actually configured, meaning that we now have to roll backwards to find which of two locations is the correct one if a user is upgrading from 18.06.

If we find the source, we read from the source, and everything is fine.

Reply:
As best I can tell, the OEM device has the hex in only one location:

```
$ binwalk -R "\x00\x18\x0a\x35\xa1" *
:
:
Scan Time:     2020-07-24 02:59:41
Target File:   /home/labby/mr16_lo_old/tmp/mtd2ro
MD5 Checksum:  c279a8cf2df7af12962c0c99444f8ad2
Signatures:    1

DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
102           0x66            Raw signature (\x00\x18\x0a\x35\xa1\x6c)
:
```
It's precisely at the beginning of the firmware so I like the idea of switching the location from OpenWrt 19 to OpenWrt 20. There is no separate configuration for the WiFi MACs.

Do you know how I might add code to migrate the MAC partition over to hex 0x66 of the original firmware so that users do not get locked out by misconfigured MACs after upgrading?

Reply:
In  8d699c6, I have moved the mac partition to 0x80000+64KB, and relinked the WiFis to pull their MAC from it as well.

Reply:
> Do you know how I might add code to migrate the MAC partition over to hex 0x66 of the original firmware so that users do not get locked out by misconfigured MACs after upgrading?

With the partition read-only, no way. And I'd actually like it to remain read-only. One could do stuff like keeping the old "mac" partition and add a small piece of code to the lower 02_network section that just sets the MAC address (via uci) if something is found there. But I don't really like that; personally, I feel like "it should be done the proper way now", but that will of course be a show stopper for all ar71xx users.

Reply:
Hmm, just looked here:
https://openwrt.org/toh/meraki/mr16#flash_layout

This clearly says that the device has a "board config" partition for 0x80000 to 0xa0000. Initial support completely ignored that; so, support for this device in ar71xx appears to be actually "broken".
Despite, certain areas are not partitioned at all.

I'd thus at least add this partition from 0x80000 to 0xa0000, and call it "config". I also do not see any reason for that strange "rootfs first" partitioning there.

Reply:
> But I don't really like that; personally, I feel like "it should be done the proper way now", but that will of course be a show stopper for all ar71xx users.

A thought:

If someone is already running ar71xx, `0xbf08000 + 0x20000` is already unlocked, because it's `rootfs`. The process for writing the MAC address would be to, during stage2 (with rootfs unmounted), do:

`mtd erase $ROOTFS_MTD; echo -ne '<\x eth mac address>' | dd of=$ROOTFS_MTD bs=1 seek=$((0x66)) count=12`

If it's possible to script this into the sysupgrade between the two versions only, then I think it would be ideal.  I just don't see an obvious way to do that.

Reply:
I'm not sure yet whether I like this. In my perception, the ar71xx support still feels "just broken", and I hesitate to add further hacks now to clean this up.
For the solution you propose, the problem is that the sysupgrade code run is that from the old version, isn't it? So, one would have to upgrade into an intermediate version before finally going into clean ath79.
In addition, if we change to denx,uimage as proposed, sysupgrade won't work anyway; so, maybe we just don't have to solve this problem, and ath79 needs to be flashed via TFTP again and fine.

Reply:
BTW, if you aim at blocking the "old" mac partition anyway, the easiest way would be to just flash ath79 and then install kmod-mtd-rw via opkg.

This would allow users with old config to remove the read-only on the config partition, and write their address to it, while theoretical "new" users with intact config partition would not be affected by the problems from the past.

Reply:
> kmod-mtd-rw

This is good. I'm also considering instances where MR16s are not physically and changing the MAC address will change the IP address. Since mtd-rw gives us userspace writes to mtd, it'd be enough to add to the configuration during sysupgrade; I can provide a script to drop into /lib/upgrade, which provides a safe way to carry the present MAC addresses into uci, and assert that the configuration will be saved (or abort and reboot at any failures); I'd post the script in the wiki article.

Reply:
For posterity:

I did sleuthing to find what happened with the original MAC address setup. Chris B realized that u-boot was pulling from `0xbfff0000`, the beginning of the ART partition, even if he did not realize Meraki was storing it in the first 64K of their own partitions (which we overwrite). Here's what `u-boot` lets us do with progmac (which does not actually write anything back to flash, probably because it had not previously been `erase`d):

```
ar7100> progmac
Usage:
progmac <serno> - Program the MAC addresses
ar7100> progmac 00:11:22:33:44:55

First 0xff last 0xff sector size 0x10000                                                                                          255
write addr: bfff0000
```

This is the code from `build-dir/target/linux/ar71xx/files/arch/mips/ath79/mach-mr16.c` from [one of Chris B's MR16 LEDE releases](https://github.com/riptidewave93/Openwrt-MR12-MR16/archive/1.1Beta.tar.gz), showing that he uses what U-boot uses:

```
static void __init MR16_setup(void)                                             
{                                                                               
        u8 *mac = (u8 *) KSEG1ADDR(0xbfff0000);                                 
                                                                                
        /* Bring up MDIO */                                                     
        ath79_register_mdio(0,0x0);                                             
                                                                                
        /* 1GB POE Port */                                                      
        ath79_init_mac(ath79_eth0_data.mac_addr, mac, 0);           
```